tcpdump/4.9.3/tcpdump-4.9.3.patch

```diff
@@ -0,0 +1,167 @@
+diff --git a/configure.ac b/configure.ac
+index 56e2a62..dd14e33 100644
+--- a/configure.ac
++++ b/configure.ac
+@@ -864,6 +864,33 @@ AC_LBL_SOCKADDR_SA_LEN
+ 
+ AC_LBL_UNALIGNED_ACCESS
+ 
++AC_ARG_WITH(wolfssl,
++    AS_HELP_STRING([--with-wolfssl]=DIR,
++		   [use wolfSSL (located in directory DIR) @<:@default=no@:>@]),
++[
++	CPPFLAGS="$CPPFLAGS -I${withval}/include/wolfssl -I${withval}/include/"
++	LDFLAGS="$LDFLAGS -L${withval}/lib"
++	LIBS="$LIBS -lwolfssl"
++	USE_WOLFSSL=yes],
++[
++	USE_WOLFSSL=no
++])
++
++if test $USE_WOLFSSL = yes
++then
++    AC_CHECK_HEADER([wolfssl/options.h])
++    if test $ac_cv_header_wolfssl_options_h = yes
++    then
++        AC_DEFINE(HAVE_WOLFSSL, 1, [define if you're using wolfSSL])
++        AC_DEFINE(HAVE_EVP_CIPHER_CTX_NEW, 1, [])
++        AC_DEFINE(HAVE_EVP_CIPHERINIT_EX, 1, [])
++        AC_DEFINE(HAVE_LIBCRYPTO, 1, [])
++        AC_DEFINE(HAVE_OPENSSL_EVP_H, 1, [])
++    else
++        AC_MSG_ERROR([Unable to find wolfssl.])
++    fi
++fi
++
+ # Check for OpenSSL/libressl libcrypto
+ AC_MSG_CHECKING(whether to use OpenSSL/libressl libcrypto)
+ # Specify location for both includes and libraries.
+@@ -872,7 +899,11 @@ AC_ARG_WITH(crypto,
+     AS_HELP_STRING([--with-crypto]@<:@=DIR@:>@,
+ 		   [use OpenSSL/libressl libcrypto (located in directory DIR, if specified) @<:@default=yes, if available@:>@]),
+ [
+-	if test $withval = no
++	if test $USE_WOLFSSL = yes
++	then
++		want_libcrypto=no
++		AC_MSG_RESULT(no)
++	elif test $withval = no
+ 	then
+ 		# User doesn't want to link with libcrypto.
+ 		want_libcrypto=no
+@@ -898,12 +929,18 @@ AC_ARG_WITH(crypto,
+ 		LIBS="-L$withval/lib $LIBS"
+ 	fi
+ ],[
++	if test $USE_WOLFSSL = yes
++	then
++		want_libcrypto=no
++		AC_MSG_RESULT(no)
+ 	#
+ 	# Use libcrypto if it's present, otherwise don't; no directory
+ 	# was specified.
+ 	#
+-	want_libcrypto=ifavailable
+-	AC_MSG_RESULT([yes, if available])
++	else
++		want_libcrypto=ifavailable
++		AC_MSG_RESULT([yes, if available])
++	fi
+ ])
+ if test "$want_libcrypto" != "no"; then
+ 	#
+diff --git a/print-esp.c b/print-esp.c
+index 6fabff1..71166bf 100644
+--- a/print-esp.c
++++ b/print-esp.c
+@@ -37,6 +37,10 @@
+  * is the simplest way of handling the dependency.
+  */
+ #ifdef HAVE_LIBCRYPTO
++#ifdef HAVE_WOLFSSL
++#include <wolfssl/options.h>
++#include <openssl/ssl.h>
++#endif
+ #ifdef HAVE_OPENSSL_EVP_H
+ #include <openssl/evp.h>
+ #else
+@@ -259,7 +263,13 @@ int esp_print_decrypt_buffer_by_ikev2(netdissect_options *ndo,
+ 		EVP_CIPHER_CTX_free(ctx);
+ 		(*ndo->ndo_error)(ndo, "can't allocate memory for decryption buffer");
+ 	}
++#ifdef HAVE_WOLFSSL
++	/* wolfSSL expects the length passed in to include the padding, so we use
```

Comment:
This seems very surprising that it is different.  Is the OpenSSL usage wrong perhaps?
Could the OpenSSL usage be the same?

Reply:
That's definitely possible. I'll circle back on the failing unit test that led to this change and see what I can find. It would be great if the usage could be the same for both OpenSSL and wolfSSL.

Reply:
@mcr, my conclusion is that yes, the OpenSSL usage could be the same. I made the following change to tcpdump (without any of the wolfSSL changes, just this one change) and all the tests pass:

```
diff --git a/print-esp.c b/print-esp.c
index 6fabff1..cdeca9d 100644
--- a/print-esp.c
+++ b/print-esp.c
@@ -259,7 +259,7 @@ int esp_print_decrypt_buffer_by_ikev2(netdissect_options *ndo,
                EVP_CIPHER_CTX_free(ctx);
                (*ndo->ndo_error)(ndo, "can't allocate memory for decryption buffer");
        }
-       EVP_Cipher(ctx, output_buffer, input_buffer, len);
+       EVP_Cipher(ctx, output_buffer, input_buffer, buffer_size);
        EVP_CIPHER_CTX_free(ctx);
 
        /*
```

The test that was failing with the old call for the wolfSSL case was ikev2pI2. If we take a look at the OpenSSL documentation for `EVP_Cipher`, we find this:

> Encrypts or decrypts a maximum inl amount of bytes from in and leaves the result in out. For legacy ciphers - If the cipher doesn't have the flag EVP_CIPH_FLAG_CUSTOM_CIPHER set, then inl must be a multiple of EVP_CIPHER_get_block_size(). If it isn't, the result is undefined. If the cipher has that flag set, then inl can be any size.

So, there are some OpenSSL ciphers that will _require_ that the input buffer length be `buffer_size`, while others will will permit `len`. Thus, I think the change above is a good one, as it should cover all ciphers.

I should also add this bit from the same documentation:

> Due to the constraints of the API contract of this function it shouldn't be used in applications, please consider using EVP_CipherUpdate() and EVP_CipherFinal_ex() instead.

It might be even better to migrate to using `EVP_CipherUpdate` and `EVP_CipherFinal_ex`, as they advise. wolfSSL has support for these functions, too, so from the wolfSSL perspective, either way is fine.